// baseball/Services/Animator.js
import { Mathinator } from 'baseball/services/_services';

var Animator = function() {
    this.init();
};
Animator.TweenMax = {};
Animator.prototype = {
    identifier : 'Animator',
    constructor : Animator,
    console : false,
    TweenMax : {},
    init : function() {

    },
    loadTweenMax : function() {
        if (this.console || typeof window !== 'object') {
            Animator.TweenMax = {
                'set': function() {},
                'to': function() {},
                'from': function() {},
                killAll: function() {}
            }
        } else {
            Animator.TweenMax = window.TweenMax;
        }
        return Animator.TweenMax;
    },
    TIME_FROM_SET : 2300, //ms
    TIME_FROM_WINDUP : 3600, //ms
    HOLD_UP_ALLOWANCE : 0.75, // seconds
    pitchTarget : null,
    pitchBreak : null,
    /**
     * this is called from the scope context
     * @param callback
     */
    updateFlightPath: function(callback) {
        if (Animator.console) return;
        var TweenMax = Animator.loadTweenMax();
        TweenMax.killAll();
        var $scope = this,
            game = $scope.y,
            top = 200-game.pitchTarget.y,
            left = game.pitchTarget.x,
            breakTop = 200-game.pitchInFlight.y,
            breakLeft = game.pitchInFlight.x,
            $baseballs = $('.baseball'),
            flightSpeed = 1.3 - 0.6*(game.pitchInFlight.velocity + 300)/400,
            originTop = 50,
            originLeft = 110 + (game.pitcher.throws == 'left' ? 20 : -20);
        var pitch = this.pitchTarget = $('.main-area .target .baseball.pitch'),
            henka = this.pitchBreak = $('.main-area .target .baseball.break'),
            quarter = flightSpeed/4;

        var pitchTransition = Mathinator.pitchTransition(top, left, originTop, originLeft, quarter, 12, 4),
            targetTransition = Mathinator.pitchTransition(top, left, originTop, originLeft, quarter, 10, 3);

        var transitions = [
            pitchTransition(0, 0),
            pitchTransition(10, 0),
            pitchTransition(30, 1),
            pitchTransition(50, 2),

            targetTransition(100, 3),
            pitchTransition(100, 3, breakTop, breakLeft)
        ];

        //var horizontalBreak = (60 - Math.abs(game.pitchTarget.x - game.pitchInFlight.x))/10;
        //$('.baseball').addClass('spin');
        //$('.baseball').css('animation', 'spin '+horizontalBreak+'s 5 0s linear');

        TweenMax.set([pitch, henka], transitions[0]);
        TweenMax.to([pitch, henka], quarter, transitions[1]);
        TweenMax.to([pitch, henka], quarter, transitions[2]);
        TweenMax.to([pitch, henka], quarter, transitions[3]);
        TweenMax.to(pitch, quarter, transitions[4]);
        TweenMax.to(henka, quarter, transitions[5]);

        $scope.lastTimeout = setTimeout(function() {
            $scope.allowInput = true;
            if (typeof callback == 'function') {
                callback();
                $scope.$apply();
            }
        }, flightSpeed*1000);

        if (!game.pitchInFlight.x) {
            $baseballs.addClass('hide');
        } else {
            if (game.humanBatting() && Math.random()*180 > game.batter.skill.offense.eye) {
                $('.baseball.break').addClass('hide');
            } else {
                $('.baseball.break').removeClass('hide');
            }
            $('.baseball.pitch').removeClass('hide');
        }

        if ($scope.y.humanBatting() && !$scope.y.humanPitching()) {
            $scope.holdUpTimeouts.push(setTimeout(function() {
                $scope.holdUp();
            }, (flightSpeed + Animator.HOLD_UP_ALLOWANCE) * 1000));
        }
    },
    animateFieldingTrajectory: function(game) {
        if (Animator.console) return game.swingResult;
        this.init();
        var TweenMax = Animator.loadTweenMax();
        var ball = $('.splay-indicator-ball');
        TweenMax.killAll();
        var result = game.swingResult;

        var linearApproximateDragScalar = {
            distance: 1,
            apexHeight: 0.57,
            airTime: 0.96
        };

        var angle = result.flyAngle,
            distance = Math.abs(result.travelDistance),
            scalar = result.travelDistance < 0 ? -1 : 1;

        Mathinator.memory.bounding = angle < 0;
        angle = 1 + Math.abs(angle);
        if (angle > 90) angle = 180 - angle;

        var velocity = linearApproximateDragScalar.distance * Math.sqrt(9.81 * distance / Math.sin(2*Math.PI*angle/180));
        var velocityVerticalComponent = Math.sin(Mathinator.RADIAN * angle) * velocity;
        var apexHeight = velocityVerticalComponent*velocityVerticalComponent/(2*9.81) * linearApproximateDragScalar.apexHeight;
        var airTime = 1.5 * Math.sqrt(2*apexHeight/9.81) * linearApproximateDragScalar.airTime; // 2x freefall equation

        //log('angle', angle, 'vel', velocity, 'apex', apexHeight, 'air', airTime, 'dist', result.travelDistance);
        var quarter = airTime/4;
        var mathinator = new Mathinator();
        var transitions = [
            mathinator.transitionalTrajectory(0, quarter, 0, apexHeight, scalar * distance, result.splay),
            mathinator.transitionalTrajectory(25, quarter, 0),
            mathinator.transitionalTrajectory(50, quarter, 1),
            mathinator.transitionalTrajectory(75, quarter, 2),
            mathinator.transitionalTrajectory(100, quarter, 3)
        ];
        TweenMax.set(ball, transitions[0]);
        TweenMax.to(ball, quarter, transitions[1]);
        TweenMax.to(ball, quarter, transitions[2]);
        TweenMax.to(ball, quarter, transitions[3]);
        TweenMax.to(ball, quarter, transitions[4]);

        ball = $('.indicator.baseball.break').removeClass('hide').show();
        var time = quarter/2;
        transitions = [
            mathinator.transitionalCatcherPerspectiveTrajectory(0, time, 0, apexHeight, scalar * distance,
                result.splay, game.pitchInFlight),
            mathinator.transitionalCatcherPerspectiveTrajectory(12.5,   time * 0.75, 0),
            mathinator.transitionalCatcherPerspectiveTrajectory(25,     time * 0.80, 1),
            mathinator.transitionalCatcherPerspectiveTrajectory(37.5,   time * 0.85, 2),
            mathinator.transitionalCatcherPerspectiveTrajectory(50,     time * 0.90, 3),
            mathinator.transitionalCatcherPerspectiveTrajectory(62.5,   time * 0.95, 4),
            mathinator.transitionalCatcherPerspectiveTrajectory(75,     time, 5),
            mathinator.transitionalCatcherPerspectiveTrajectory(87.5,   time, 6),
            mathinator.transitionalCatcherPerspectiveTrajectory(100,    time, 7)
        ];
        TweenMax.set(ball, transitions[0]);
        TweenMax.to(ball, time, transitions[1]);
        TweenMax.to(ball, time, transitions[2]);
        TweenMax.to(ball, time, transitions[3]);
        TweenMax.to(ball, time, transitions[4]);
        TweenMax.to(ball, time, transitions[5]);
        TweenMax.to(ball, time, transitions[6]);
        TweenMax.to(ball, time, transitions[7]);
        TweenMax.to(ball, time, transitions[8]);

        setTimeout(function() {
            // hack
            $('.indicator.baseball.break').removeClass('hide').show();
        }, 50);

        return game.swingResult;
    }
};

for (var fn in Animator.prototype) {
    if (Animator.prototype.hasOwnProperty(fn)) {
        Animator[fn] = Animator.prototype[fn];
    }
}

export { Animator }
