// baseball/Model/Game.js
import { Field } from 'baseball/Model/Field';
import { Team } from 'baseball/Model/Team';
import { Umpire } from 'baseball/Model/Umpire';
import { Log } from 'baseball/Utility/Log';

import { helper, text } from 'baseball/Utility/_utils';

import { Animator, Distribution, Mathinator, Iterator } from 'baseball/Services/_services';

var Game = function(m) {
    this.init(m);
};

Game.prototype = {
    constructor : Game,
    gamesIntoSeason : 0,
    humanControl : 'home', //home, away, both, none
    console : false,
    quickMode : true,
    debug : [],
    pitcher : {}, // Player&
    batter : {}, // Player&
    init : function(m) {
        if (m) text.mode = m;
        this.gamesIntoSeason = 1 + Math.floor(Math.random()*142);
        this.field = new Field(this);
        this.teams.away = new Team(this);
        this.teams.home = new Team(this);
        this.log = new Log();
        this.log.game = this;
        this.debug = [];
        this.helper = helper;
        while (this.teams.away.name == this.teams.home.name) {
            this.teams.away.pickName();
        }
        this.umpire = new Umpire(this);
        if (this.humanPitching()) {
            this.stage = 'pitch';
        }
    },
    getInning : function() {
        return text.mode == 'n' ? (this.inning + (this.half == 'top' ? 'オモテ' : 'ウラ')) : this.half.toUpperCase() + ' ' + this.inning;
    },
    humanBatting : function() {
        if (this.humanControl == 'none') return false;
        switch (this.half) {
            case 'top':
                return this.humanControl == 'both' || this.humanControl == 'away';
            case 'bottom':
                return this.humanControl == 'both' || this.humanControl == 'home';
        }
    },
    humanPitching : function() {
        if (this.humanControl == 'none') return false;
        switch (this.half) {
            case 'top':
                return this.humanControl == 'both' || this.humanControl == 'home';
            case 'bottom':
                return this.humanControl == 'both' || this.humanControl == 'away';
        }
    },
    end : function() {
        this.stage = 'end';
        var e, n;
        e = this.tally.home.R > this.tally.away.R ? 'Home team wins!' :
            (this.tally.home.R == this.tally.away.R ? 'You tied. Yes, you can do that.' : 'Visitors win!');
        n = this.tally.home.R > this.tally.away.R ? this.teams.home.getName()+'の勝利' :
            (this.tally.home.R == this.tally.away.R ? '引き分け' : this.teams.away.getName()+'の勝利');
        if (this.tally.home.R > this.tally.away.R) {
            this.teams.home.positions.pitcher.stats.pitching.W++;
            this.teams.away.positions.pitcher.stats.pitching.L++;
        } else if (this.tally.home.R < this.tally.away.R) {
            this.teams.home.positions.pitcher.stats.pitching.L++;
            this.teams.away.positions.pitcher.stats.pitching.W++;
        }
        this.log.note(e, n);
        this.log.note('Reload to play again', 'リロるは次の試合へ');
    },
    stage : 'pitch', //pitch, swing
    simulateInput : function(callback) {
        if (this.stage == 'end') {
            return;
        }
        if (this.stage == 'pitch') {
            this.autoPitch(callback);
        } else if (this.stage == 'swing') {
            if (typeof this.pitchTarget != 'object') {
                this.pitchTarget = {x: 100, y: 100};
            }
            this.autoSwing(this.pitchTarget.x, this.pitchTarget.y, callback);
        }
    },
    simulatePitchAndSwing : function(callback) {
        if (this.stage == 'end') {
            return;
        }
        this.autoPitch(callback);
        var giraffe = this;
        setTimeout(function() {
            if (typeof giraffe.pitchTarget != 'object') {
                giraffe.pitchTarget = {x: 100, y: 100};
            }
            giraffe.autoSwing(giraffe.pitchTarget.x, giraffe.pitchTarget.y, function(callback) {callback();});
        }, giraffe.field.hasRunnersOn() ? Animator.TIME_FROM_SET + 1200 : Animator.TIME_FROM_WINDUP + 1200);
    },
    /**
     * generically receive click input and decide what to do
     * @param x
     * @param y
     * @param callback
     */
    receiveInput : function(x, y, callback) {
        if (this.humanControl == 'none') {
            return;
        }
        if (this.stage == 'end') {
            return;
        }
        if (this.stage == 'pitch' && this.humanPitching()) {
            this.thePitch(x, y, callback);
        } else if (this.stage == 'swing'  && this.humanBatting()) {
            this.theSwing(x, y, callback);
        }
    },
    autoPitchSelect : function() {
        var pitchName = this.helper.selectRandomPitch();
        while (!this.pitcher.pitching.hasOwnProperty(pitchName)) {
            pitchName = this.helper.selectRandomPitch();
        }
        var pitch = this.pitcher.pitching[pitchName];
        pitch.name = pitchName;
        this.pitchInFlight = pitch;
    },
    autoPitch : function(callback) {
        var pitcher = this.pitcher,
            giraffe = this;
        if (this.stage == 'pitch') {
            pitcher.windingUp = true;
            if (!this.console) {
                $('.baseball').addClass('hide');
                var windup = $('.windup');
                windup.css('width', '100%');
            }
            this.autoPitchSelect();
            var pitch = Distribution.pitchLocation(),
                x = pitch.x,
                y = pitch.y;
            if (this.quickMode) {
                this.thePitch(x, y, callback);
            } else {
                windup.animate({width: 0}, this.field.hasRunnersOn() ? Animator.TIME_FROM_SET : Animator.TIME_FROM_WINDUP, function() {
                    !giraffe.console && $('.baseball.pitch').removeClass('hide');
                    giraffe.thePitch(x, y, callback);
                    pitcher.windingUp = false;
                });
            }
        }
    },
    autoSwing : function(deceptiveX, deceptiveY, callback) {
        var giraffe = this;
        var bonus = this.batter.eye.bonus || 0,
            eye = this.batter.skill.offense.eye + 6*(this.umpire.count.balls + this.umpire.count.strikes) + bonus,
            convergence, convergenceSum;

        var x = Distribution.centralizedNumber(),
            y = Distribution.centralizedNumber();
        if (100*Math.random() < eye) {
            convergence = 1.35 * 5*eye/100;
            convergenceSum = 1 + convergence;
            deceptiveX = this.pitchInFlight.x;
            deceptiveY = this.pitchInFlight.y;
        } else {
            convergence = 1.35 * 2*eye/100;
            convergenceSum = 1 + convergence;
        }
        x = (deceptiveX*(convergence) + x)/convergenceSum;
        y = (deceptiveY*(convergence) + y)/convergenceSum;

        this.swingResult.x = Distribution.cpuSwing(x, this.pitchInFlight.x, eye);
        this.swingResult.y = Distribution.cpuSwing(y, this.pitchInFlight.y, eye);

        var swingProbability = Distribution.swingLikelihood(eye, x, y, this.umpire);
        if (swingProbability < 100*Math.random()) {
            x = -20;
        }

        callback(function() {
            giraffe.theSwing(x, y);
        });
    },
    opponentConnected : false,
    waitingCallback : function() {},
    awaitPitch : function(callback, swingResult) {
        if (this.opponentConnected) {
            this.waitingCallback = callback;
            this.opponentService.emitSwing(swingResult);
        } else {
            this.autoPitch(callback);
        }
    },
    awaitSwing : function(x, y, callback, pitchInFlight, pitchTarget) {
        if (this.opponentConnected) {
            this.waitingCallback = callback;
            this.opponentService.emitPitch({
                inFlight: pitchInFlight,
                target: pitchTarget
            });
        } else {
            this.autoSwing(x, y, callback);
        }
    },
    thePitch : function(x, y, callback, override) {
        if (this.stage == 'pitch') {
            if (override) {
                this.pitchInFlight = override.inFlight;
                this.pitchTarget = override.target;
                callback = this.waitingCallback;
            } else {
                this.pitcher.fatigue++;
                this.pitchTarget.x = x;
                this.pitchTarget.y = y;

                this.pitchInFlight.breakDirection = this.helper.pitchDefinitions[this.pitchInFlight.name].slice(0, 2);
                this.battersEye = text.getBattersEye(this);

                var control = Math.floor(this.pitchInFlight.control - this.pitcher.fatigue/2);
                this.pitchTarget.x = Distribution.pitchControl(this.pitchTarget.x, control);
                this.pitchTarget.y = Distribution.pitchControl(this.pitchTarget.y, control);

                if (this.pitcher.throws == 'right') this.pitchInFlight.breakDirection[0] *= -1;

                var breakEffect = Distribution.breakEffect(this.pitchInFlight, this.pitcher, this.pitchTarget.x, this.pitchTarget.y);

                this.pitchInFlight.x = breakEffect.x;
                this.pitchInFlight.y = breakEffect.y;
            }
            this.log.notePitch(this.pitchInFlight, this.batter);

            this.stage = 'swing';
            if (this.humanControl != 'none' && (this.humanControl == 'both' || this.humanBatting())) {
                callback();
            } else {
                this.awaitSwing(x, y, callback, this.pitchInFlight, this.pitchTarget);
            }
        }
    },
    battersEye : {
        e: '',
        n: ''
    },
    theSwing : function(x, y, callback, override) {
        if (this.stage == 'swing') {
            if (override) {
                this.swingResult = override;
                callback = this.waitingCallback;
            } else {
                this.swingResult = {};
                var bonus = this.batter.eye.bonus || 0,
                    eye = this.batter.skill.offense.eye + 6*(this.umpire.count.balls + this.umpire.count.strikes) + bonus;

                if (x >= 0 && x <= 200) {
                    this.batter.fatigue++;

                    this.swingResult.x = x - this.pitchInFlight.x;
                    this.swingResult.y = y - this.pitchInFlight.y;
                    this.swingResult.angle = this.setBatAngle();

                    var recalculation = Mathinator.getAngularOffset(this.swingResult, this.swingResult.angle);
                    var precision = Distribution.swing(eye);

                    this.swingResult.x = Math.abs(recalculation.x) > 20 ? recalculation.x * precision : recalculation.x;
                    this.swingResult.y = -5 + (recalculation.y < 0 ? recalculation.y * precision : recalculation.y);

                    //log(recalculation.y, precision);

                    this.swingResult.looking = false;
                    if (Math.abs(this.swingResult.x) < 60 && Math.abs(this.swingResult.y) < 35) {
                        this.swingResult.contact = true;
                        this.swingResult = this.field.determineSwingContactResult(this.swingResult);
                        // log(this.swingResult.flyAngle, Math.floor(this.swingResult.x), Math.floor(this.swingResult.y));
                        this.debug.push(this.swingResult);
                    } else {
                        this.swingResult.contact = false;
                    }
                } else {
                    this.swingResult.strike = this.pitchInFlight.x > 50 && this.pitchInFlight.x < 150
                        && this.pitchInFlight.y > 35 && this.pitchInFlight.y < 165;
                    this.batter.eye.bonus = Math.max(0, eye -
                        Math.sqrt(Math.pow(this.batter.eye.x - this.pitchInFlight.x, 2) + Math.pow(this.batter.eye.y - this.pitchInFlight.y, 2)) * 1.5);
                    this.swingResult.contact = false;
                    this.swingResult.looking = true;
                    this.batter.eye.x = this.pitchInFlight.x;
                    this.batter.eye.y = this.pitchInFlight.y;
                }
            }
            this.log.noteSwing(this.swingResult);
            this.stage = 'pitch';

            var half = this.half;
            var result = this.swingResult
            this.umpire.makeCall();
            emit = false;
            if (half != this.half) {
                callback = this.startOpponentPitching;
                var emit = !override;
            }

            if (typeof callback == 'function') {
                if (this.humanControl != 'none' && (this.humanControl == 'both' || this.teams[this.humanControl] == this.pitcher.team)) {
                    callback();
                    if (emit) {
                        if (this.opponentService && this.opponentConnected) {
                            this.opponentService.emitSwing(result);
                        }
                    }
                } else {
                    this.awaitPitch(callback, result);
                }
            }
        }
    },
    setBatAngle : function(x, y) {
        var giraffe = this;
        var origin = {
            x: giraffe.batter.bats == 'right' ? -10 : 210,
            y: 160
        };
        var swing = {
            x: x ? x : giraffe.pitchInFlight.x + this.swingResult.x,
            y: y ? y : giraffe.pitchInFlight.y + this.swingResult.y
        };
        return Mathinator.battingAngle(origin, swing);
    },
    debugOut : function() {
        log('slugging', this.debug.filter(function(a){return a.bases == 1;}).length,
            this.debug.filter(function(a){return a.bases == 2;}).length,
            this.debug.filter(function(a){return a.bases == 3;}).length,
            this.debug.filter(function(a){return a.bases == 4;}).length
        );
        log('grounders', this.debug.filter(function(a){return !a.caught && !a.foul && a.flyAngle < -5;}).length);
        log('grounders thrown out', this.debug.filter(function(a){return !a.caught && !a.foul && a.flyAngle < -5 && a.thrownOut;}).length);
        log('weak fly hits (thrown out)', this.debug.filter(function(a){return !a.caught && !a.foul && a.flyAngle > 0 && a.thrownOut;}).length);
        log('good fly hits (not caught)', this.debug.filter(function(a){return !a.caught && !a.foul && a.flyAngle > 0 && !a.thrownOut;}).length);

        var PO = {};
        this.debug.map(function(a) {
            if (!PO[a.fielder]) {
                PO[a.fielder] = 0;
            }
            if (!a.bases && a.fielder) {
                PO[a.fielder]++;
            }
        });
        log('fielding outs', PO);

        var hitters = this.teams.away.lineup.concat(this.teams.home.lineup);
        var atBats = [];
        hitters.map(function(a) {atBats = atBats.concat(a.atBats);});

        var LO = atBats.filter(function(ab) {return ab == 'LO';}).length;
        log('line outs', LO);
        var FO = atBats.filter(function(ab) {return ab == 'FO';}).length;
        log('fly outs', FO);
        var GO = atBats.filter(function(ab) {return ab == 'GO';}).length;
        log('groundouts', GO);
        var SO = atBats.filter(function(ab) {return ab == 'SO';}).length;
        log('strikeouts', SO);
        log('fouls', this.debug.filter(function(a){return a.foul;}).length);
        log('fatigue, home vs away');
        var teams = this.teams;
        Iterator.each(this.teams.home.positions, function(key, value) {
            var position = key;
            log(key, teams.home.positions[position].fatigue, teams.away.positions[position].fatigue)
        });
    },
    toData : function() {
        var data = {};
        data.half = this.half;
        data.inning = this.inning;
        data.tally = this.tally;
        var giraffe = this;
        var players = this.teams.away.lineup.concat(this.teams.home.lineup);
        // note: bench not included
        data.field = {
            first: players.indexOf(this.field.first),
            second: players.indexOf(this.field.second),
            third: players.indexOf(this.field.third)
        };
        data.batter = players.indexOf(this.batter);
        data.deck = players.indexOf(this.deck);
        data.hole = players.indexOf(this.hole);
        data.teams = {
            home: {
                name: giraffe.teams.home.name,
                nameJ: giraffe.teams.home.nameJ
            },
            away: {
                name: giraffe.teams.away.name,
                nameJ: giraffe.teams.away.nameJ
            }
        };
        data.umpire = {
            says : giraffe.umpire.says,
            count: {
                strikes : giraffe.umpire.count.strikes,
                balls: giraffe.umpire.count.balls,
                outs: giraffe.umpire.count.outs
            }
        };
        data.players = players.map(function(player) {
            return player.serialize();
        });
        data.log = {
            pitchRecord : giraffe.log.pitchRecord,
            record : giraffe.log.record
        };
        data.gamesIntoSeason = this.gamesIntoSeason;
        return data;
    },
    fromData : function(data) {
        this.half = data.half;
        this.inning = data.inning;
        this.tally = data.tally;
        var giraffe = this;
        var players = data.players.map(function(playerJson, index) {
            var playerData = JSON.parse(playerJson);
            if (index > 8) {
                var side = 'home';
                index = index - 9;
            } else {
                side = 'away';
            }
            var player = giraffe.teams[side].positions[playerData.position];
            player.fromData(playerData);
            giraffe.teams[side].lineup[index] = player;
            player.resetStats(data.gamesIntoSeason);
            return player;
        });
        this.field.first = players[data.field.first];
        this.field.second = players[data.field.second];
        this.field.third = players[data.field.third];
        this.batter = players[data.batter];
        this.deck = players[data.deck];
        this.hole = players[data.hole];
        this.umpire.says = data.umpire.says;
        this.umpire.count = data.umpire.count;
        this.teams.away.name = data.teams.away.name;
        this.teams.away.nameJ = data.teams.away.nameJ;
        this.teams.home.name = data.teams.home.name;
        this.teams.home.nameJ = data.teams.home.nameJ;
        this.log.pitchRecord = data.log.pitchRecord;
        this.log.record = data.log.record;
        this.log.stabilizeShortRecord();
        this.gamesIntoSeason = data.gamesIntoSeason;
        return this;
    },
    startOpponentPitching : null, // late function
    pitchTarget : {x : 100, y : 100},
    pitchInFlight : {
        x : 100,
        y : 100,
        breakDirection : [0, 0],
        name : 'slider',
        velocity : 50,
        'break' : 50,
        control : 50
    },
    swingResult : {
        x : 100, //difference to pitch location
        y : 100, //difference to pitch location
        strike : false,
        foul : false,
        caught : false,
        contact : false,
        looking : true,
        bases : 0,
        fielder : 'short',
        outs : 0
    },
    playResult : {
        batter: '',
        fielder: ''
    },
    showPlayResultPanels : function(batter) {
        var batterOutcomes = {
        };
        var atBat = batter.atBats.slice(0).pop();
        var fielderOutcomes = {
        };
        var n = function() {
            var n = Math.floor(Math.random()*3);
            return n ? n : '';
        };
        this.playResult = {
            batter: 'B_placeholder' + n() || batterOutcomes[atBat] || 'batter/' + atBat,
            fielder: 'F_placeholder' + n() || fielderOutcomes[atBat] || 'fielder/' + atBat
        };
    },
    pitchSelect : function() {

    },
    field : null,
    teams : {
        away : null,
        home : null
    },
    log : null,
    half : 'top',
    inning : 1,
    scoreboard : {
        away : {
            1 : 0,
            2 : 0,
            3 : 0,
            4 : 0,
            5 : 0,
            6 : 0,
            7 : 0,
            8 : 0,
            9 : 0
        },
        home : {
            1 : 0,
            2 : 0,
            3 : 0,
            4 : 0,
            5 : 0,
            6 : 0,
            7 : 0,
            8 : 0,
            9 : 0
        }
    },
    resetTally : function() {
        this.tally = {
            away : {
                H : 0,
                R : 0,
                E : 0
            },
            home : {
                H : 0,
                R : 0,
                E : 0
            }
        };
    },
    tally : {
        away : {
            H : 0,
            R : 0,
            E : 0
        },
        home : {
            H : 0,
            R : 0,
            E : 0
        }
    }
};

export { Game }
