// baseball/Model/Team.js
import { Player } from 'baseball/Model/Player';
import { Manager } from 'baseball/Model/Manager';

import { data, text } from 'baseball/Utility/_utils';

var Team = function(game) {
    this.init(game);
};

Team.RUNNERS_DISCRETION = 'runnersDiscretion';
Team.RUNNER_GO = 'go';
Team.RUNNER_HOLD = 'hold';

Team.prototype = {
    constructor : Team,
    init : function(game) {
        this.substituted = [];
        this.pickName();
        this.lineup = [];
        this.bench = [];
        this.bullpen = [];
        this.positions = {
            pitcher : null,
            catcher : null,
            first : null,
            second : null,
            short : null,
            third : null,
            left : null,
            center : null,
            right : null
        };
        this.manager = new Manager(this);
        if (game !== 'no init') {
            this.game = game;
            for (var j = 0; j < 20; j++) {
                this.bench.push(new Player(this));
            }
            if (this.bench.length == 20) {
                this.manager.makeLineup();
            }
        }
    },
    pickName : function() {
        var teamNameIndex = Math.floor(Math.random()*data.teamNames.length);
        this.name = data.teamNames[teamNameIndex];
        this.nameJ = data.teamNamesJ[teamNameIndex];
    },
    getName : function() {
        return text.mode == 'n' ? this.nameJ : this.name;
    },
    stealAttempt : Team.RUNNERS_DISCRETION,
    lineup : [],
    positions : {},
    manager : null,
    bench : [],
    bullpen : [],
    nowBatting : 0,
    expanded : 'Player&'
};

export { Team }
